src/fal/dag.py

```diff
@@ -0,0 +1,129 @@
+
+    def __hash__(self):
+        return self.path.__hash__()
+
+    def __repr__(self):
+        return "FalScript(" + self.path.__str__() + ")"
+
+    def parse(self) -> List[str]:
+        """
+        Parses the python file to identify any dependencies
+        """
+        with open(self.path) as file:
+            lines = file.readlines()
+            ## TODO: do this search in parallel
+            dependency_names = []
+            for line in lines:
+                match = re.search('ref\("(.*)"\)', line)
+                if match:
+                    group = match.group(1)
+                    dependency_names.append(group)
+            return dependency_names    
+    
+    def addDependency(self, dependency: T):
+        if (dependency not in self.dependencies):
+            self.dependencies.append(dependency)
+
+    def exec(self, ref, context, source):
+        """
+        Executes the script
+        """
+        with open(path) as file:
+            a_script = file.read()
+            exec(
+                a_script,
+                {
+                    "ref": ref,
+                    "context": context,
+                    "source": source,
+                },
+            )
+
+@dataclass(frozen=True)
+class UniqueKey:
+    model: DbtModel
+    script_path: Path
+
+class ScriptGraphBuilder:
+    modelNameToModelLookup: Dict[str, DbtModel] = {}
+    modelToScriptLookup: Dict[str, List[Path]] = {}
+    falScripts: Dict[UniqueKey, FalScript] = {}
+
+    def __init__(self, models: List[DbtModel], keyword: str, root: str):
+        for model in models:
+            self.modelToScriptLookup[model.name] = model.get_scipts(keyword, root)
+            self.modelNameToModelLookup[model.name] = model
+        # self.scripts = _flatten(self.modelToScriptLookup.values())
+        for model_name, script_list in self.modelToScriptLookup.items():
+            for script in script_list:
+                self.recursivelySetDependencies(UniqueKey(self.modelNameToModelLookup[model_name], script))
+
+    def recursivelySetDependencies(self, key: UniqueKey):
+        if (self.falScripts.get(key) is None):
+            self.falScripts[key] = FalScript(key.model, key.script_path)
+
+        current_script = self.falScripts.get(key)
+        dependency_models = current_script.parse()
+       
+        for model_name in dependency_models:
+            model = self.modelNameToModelLookup[model_name]
+            dependency_scripts = self.modelToScriptLookup.get(model_name, [])
+
+            for dependency_scipt in dependency_scripts:
+                if (self.falScripts.get(UniqueKey(model, dependency_scipt)) is None):
+                    self.recursivelySetDependencies(UniqueKey(model, dependency_scipt))
+                current_script.addDependency(self.falScripts.get(UniqueKey(model, dependency_scipt)))
+    
+    def getValues(self) -> List[FalScript]:
+        return self.falScripts.values()
+
+
+class ScriptGraph:
```

Comment:
Still have to review this class more

Reply:
I am using this algorithm by the way, helps with the review 

https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm